### PR TITLE
feat: Implied 'member' roles for site and organization

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

This modifies the sql query that fetchs roles for authorization to include 'member' and 'organization-member' automatically. This implication means you can never be a user of the platform, and not be a member. And if you are in an organization, you are a organization-member of that org.

This just makes sense, as a user with no roles doesn't make much sense. This just makes sure a user with "no roles" never exists. An organization-member always implies they have their member role. It was weird you could have a user/member with no roles.

# UI implications

'member' is never shown on the ui anymore. TBD if that is good UX

## As an admin

Screenshot from 2022-05-31 09-56-06

## As a member

Screenshot from 2022-05-31 09-55-03

# UX implications

You can never assign or remove any 'member' roles. They are automatic

### CODE REVIEW

File: coderd/httpmw/apikey.go [hunk 1]
```diff
@@ -31,6 +31,19 @@ func APIKey(r *http.Request) database.APIKey {
 	return apiKey
 }
 
+// User roles are the 'subject' field of Authorize()
+type userRolesKey struct{}
+
+// UserAuthorizationRoles returns the roles used for authorization.
+// Comes from the ExtractAPIKey handler.
+func UserAuthorizationRoles(r *http.Request) database.GetAuthorizationUserRolesRow {
```
Comment (johnstcn) [inline coderd/httpmw/apikey.go:39]: nit: `httpmw.UserAuthorizationRoles` vs `database.GetAuthorizationUserRoles` -- pick one or othe other, else it'll be super annoying to find later!

File: coderd/rbac/role.go [hunk 2]
```diff
@@ -17,7 +17,9 @@ type Permission struct {
 // Users of this package should instead **only** use the role names, and
 // this package will expand the role names into their json payloads.
 type Role struct {
-	Name        string       `json:"name"`
+	Name string `json:"name"`
+	// DisplayName is used for UI purposes. If the role has no display name,
+	// that means the UI should never display it.
 	DisplayName string       `json:"display_name"`
```
Comment (johnstcn) [inline coderd/rbac/role.go:23]: Can we just not send this role to the UI at all? I think the whole `org-member:` thing is just the Rego leaking out a bit.

**Review by johnstcn — APPROVED** [review]
I was worried initially that this concept of "implied roles" makes things too complicated.

As the whole `organization-member:<org_id>` is just an internal implementation detail of the Rego, I had wondered if we could just stuff that whole issue inside the `rbac` package and keep it locked away in there.

But we still need to handle cases where users are a member of an org but do not have any additional roles in the org, just their membership.

We don't want to introduce any `database` dependencies in the `rbac` package at all, so the `rbac` package has no real way of just "magically" fetching the user org memberships.

I then thought about passing in the organization IDs of which the subject is a member into `authz.Authorize` as another argument, but then you can get into weird situations where a user does not have membership of an org but still has roles in that org.

So the solution here makes sense to me.

**Comment by Emyrk** [comment]
> I was worried initially that this concept of "implied roles" makes things too complicated.
> 
> As the whole `organization-member:<org_id>` is just an internal implementation detail of the Rego, I had wondered if we could just stuff that whole issue inside the `rbac` package and keep it locked away in there.
> 
> But we still need to handle cases where users are a member of an org but do not have any additional roles in the org, just their membership.
> 
> We don't want to introduce any `database` dependencies in the `rbac` package at all, so the `rbac` package has no real way of just "magically" fetching the user org memberships.
> 
> I then thought about passing in the organization IDs of which the subject is a member into `authz.Authorize` as another argument, but then you can get into weird situations where a user does not have membership of an org but still has roles in that org.
> 
> So the solution here makes sense to me.

Pretty much had a similar thought process.

File: coderd/httpmw/apikey.go (hunk 1 above)
Comment (Emyrk) [reply coderd/httpmw/apikey.go:39]: :+1:

File: coderd/rbac/role.go (hunk 2 above)
Comment (johnstcn) [reply coderd/rbac/role.go:23]: Ah gotcha.

**Pull request merged**